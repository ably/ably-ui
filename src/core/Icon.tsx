import React, { CSSProperties, useCallback, useId, useMemo } from "react";

import {
  defaultIconSecondaryColor,
  getHeroicon,
  setUniqueIds,
  toPascalCase,
} from "./Icon/utils";
import * as IconComponents from "./Icon/components";
import { IconName, IconSize } from "./Icon/types";
import { ColorClass, ColorThemeSet } from "./styles/colors/types";
import { convertTailwindClassToVar } from "./styles/colors/utils";
import cn from "./utils/cn";

export type IconProps = {
  name: IconName;
  size?: IconSize;
  color?: ColorClass | ColorThemeSet;
  secondaryColor?: ColorClass | ColorThemeSet;
  additionalCSS?: string;
  variant?: "outline" | "solid" | "micro" | "mini";
};

// Helper function to get a heroicon component from the static map
const getHeroiconComponent = (
  iconName: string,
): React.ComponentType<React.SVGProps<SVGSVGElement>> | null => {
  // Parse the icon name to extract heroicon name and variant
  // Expected format: icon-gui-{heroicon-name}-{variant}
  const match = iconName.match(/^icon-gui-(.+)-(outline|solid|mini|micro)$/);
  if (!match) {
    return null;
  }

  const [, heroiconName, variant] = match;
  const component = getHeroicon(heroiconName, variant);
  return component;
};

const Icon = ({
  name,
  size = "0.75rem",
  color,
  secondaryColor,
  additionalCSS = "",
  variant,
  ...additionalAttributes
}: IconProps) => {
  const uniqueId = useId();

  const [lightSecondaryColor, darkSecondaryColor] = (
    secondaryColor ?? ""
  ).split(" dark:") as [ColorClass, ColorClass | undefined];

  const iconName =
    variant && !name.endsWith(`-${variant}`) ? `${name}-${variant}` : name;

  // Get the local icon component (memoized)
  const LocalIconComponent = useMemo(():
    | React.ComponentType<React.SVGProps<SVGSVGElement>>
    | undefined => {
    return IconComponents[
      toPascalCase(iconName) as keyof typeof IconComponents
    ];
  }, [iconName]);

  // Get the heroicon component only if no local component is found (memoized)
  const HeroiconComponent = useMemo(() => {
    return LocalIconComponent ? null : getHeroiconComponent(iconName);
  }, [LocalIconComponent, iconName]);

  const setUniqueIdsRef = useCallback(
    (el: SVGSVGElement | null) => {
      setUniqueIds(el, uniqueId);
    },
    [uniqueId],
  );

  const iconSvg = useCallback(
    (secondaryColor?: ColorClass, isDark?: boolean, isThemed?: boolean) => {
      if (LocalIconComponent) {
        let secondaryColorValue;
        if (secondaryColor) {
          secondaryColorValue = convertTailwindClassToVar(secondaryColor);
        } else if (defaultIconSecondaryColor(name)) {
          secondaryColorValue = convertTailwindClassToVar(
            defaultIconSecondaryColor(name),
          );
        }

        return (
          <LocalIconComponent
            className={cn(
              { [`${color}`]: color },
              { [`${additionalCSS}`]: additionalCSS },
              {
                "hidden dark:block": secondaryColor && !isDark && isThemed,
              },
              {
                "dark:hidden": secondaryColor && isDark && isThemed,
              },
            )}
            style={
              {
                width: size,
                height: size,
                ...(secondaryColorValue && {
                  "--ui-icon-secondary-color": secondaryColorValue,
                }),
              } as CSSProperties
            }
            {...additionalAttributes}
            ref={setUniqueIdsRef}
          />
        );
      }

      if (HeroiconComponent) {
        return (
          <HeroiconComponent
            className={cn(
              { [`${color}`]: color },
              { [`${additionalCSS}`]: additionalCSS },
            )}
            style={{
              width: size,
              height: size,
            }}
            {...additionalAttributes}
          />
        );
      }
    },
    [
      LocalIconComponent,
      HeroiconComponent,
      iconName,
      size,
      color,
      additionalCSS,
      lightSecondaryColor,
      darkSecondaryColor,
      name,
    ],
  );

  if (lightSecondaryColor && darkSecondaryColor) {
    return (
      <>
        {iconSvg(lightSecondaryColor, false, !!darkSecondaryColor)}
        {iconSvg(darkSecondaryColor, true, true)}
      </>
    );
  } else if (lightSecondaryColor) {
    return iconSvg(lightSecondaryColor, false, !!darkSecondaryColor);
  } else {
    return iconSvg();
  }
};

export default Icon;
