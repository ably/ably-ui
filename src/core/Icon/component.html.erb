<%= tag.svg class: "#{color} #{additional_css}", style: "width: #{size}; height: #{size}", **additional_attributes do %>
  <%= tag.use "xlink:href": "#sprite-#{name}" %>
<% end %>
