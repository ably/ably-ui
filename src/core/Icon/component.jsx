import React from "react";
import T from "prop-types";

const Icon = ({ name, size = "0.75rem", color = "", additionalCSS = "", ...additionalAttributes }) => {
  return (
    <svg className={`${color} ${additionalCSS}`} style={{ width: size, height: size }} {...additionalAttributes}>
      <use xlinkHref={`#sprite-${name}`} />
    </svg>
  );
};

Icon.propTypes = {
  name: T.string.isRequired,
  size: T.string,
  color: T.string,
  additionalCSS: T.string,
};

export default Icon;
