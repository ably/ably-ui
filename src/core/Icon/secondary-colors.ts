import { ColorClass } from "../styles/colors/types";
import { IconName } from "./types";

export const defaultIconSecondaryColor = (name: IconName) => {
  const colorLookup = {
    "icon-gui-check-circled-fill": "text-white",
  } as Record<IconName, ColorClass>;

  return colorLookup[name];
};
