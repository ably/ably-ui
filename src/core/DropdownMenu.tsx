import React, {
  createContext,
  useContext,
  useState,
  useEffect,
  useRef,
  ReactNode,
  Dispatch,
} from "react";
import Icon from "./Icon";

const DropdownMenuContext = createContext<{
  isOpen: boolean;
  setOpen: Dispatch<React.SetStateAction<boolean>>;
}>({
  isOpen: false,
  setOpen: () => {},
});

type DropdownMenuProps = {
  children: ReactNode;
};

type TriggerProps = {
  children: ReactNode;
  additionalTriggerCSS?: string;
};

type ContentProps = {
  children: ReactNode;
  anchorPosition?: string;
  additionalContentCSS?: string;
};

type LinkProps = {
  url: string;
  title: string;
  subtitle: string;
  iconName: string;
  children: ReactNode;
};

const DropdownMenu = ({ children }: DropdownMenuProps) => {
  const [isOpen, setOpen] = useState(false);
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const clickHandler = (e: MouseEvent) => {
      if (ref.current?.contains(e.target as Node)) return;
      setOpen(false);
    };

    document.addEventListener("click", clickHandler);

    return () => {
      document.removeEventListener("click", clickHandler);
    };
  }, []);

  return (
    <DropdownMenuContext.Provider value={{ isOpen, setOpen }}>
      <div id="dropdown-menu" className="relative" ref={ref}>
        {children}
      </div>
    </DropdownMenuContext.Provider>
  );
};

const Trigger = ({ children, additionalTriggerCSS = "" }: TriggerProps) => {
  const { isOpen, setOpen } = useContext(DropdownMenuContext);

  return (
    <button
      id="menu-trigger"
      onClick={() => setOpen(!isOpen)}
      className={`${additionalTriggerCSS} flex items-center p-8 ml-8 group hover:text-gui-hover hover:bg-light-grey active:text-gui-active focus:text-gui-focus focus:outline-none rounded-lg`}
    >
      <span className="leading-normal">{children}</span>
      <Icon
        name="icon-gui-disclosure-arrow"
        color="text-cool-black"
        size="1.25rem"
        additionalCSS="transform rotate-90 group-hover:text-gui-hover group-active:text-gui-active group-focus:text-gui-focus"
      />
    </button>
  );
};

const Content = ({
  children,
  anchorPosition = "right",
  additionalContentCSS,
}: ContentProps) => {
  const { isOpen } = useContext(DropdownMenuContext);

  if (!isOpen) {
    return null;
  }

  const anchorPositionClasses =
    anchorPosition === "right" ? "right-0" : "left-0";

  return (
    <div
      id="menu-content"
      className={`${additionalContentCSS} absolute p-8 z-10 border border-mid-grey bg-white rounded shadow-container ${anchorPositionClasses}`}
      style={{ minWidth: 275, top: 44 }}
    >
      {children}
    </div>
  );
};

const Link = ({ url, title, subtitle, iconName, children }: LinkProps) => {
  return (
    <a
      href={url}
      className="menu-link group block p-8 hover:bg-light-grey hover:text-cool-black rounded"
    >
      <p className="mb-4">
        {title}
        {iconName ? (
          <Icon
            name={iconName}
            size="1rem"
            color="text-cool-black"
            additionalCSS="align-middle ml-8 relative -top-1 -left-4"
          />
        ) : null}
      </p>
      {subtitle ? <p className="ui-text-p3 mb-16">{subtitle}</p> : null}
      {children}
    </a>
  );
};

DropdownMenu.Trigger = Trigger;
DropdownMenu.Content = Content;
DropdownMenu.Link = Link;
export default DropdownMenu;
