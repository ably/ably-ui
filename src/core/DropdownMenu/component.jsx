import React, { createContext, useContext, useState, useEffect, useRef } from "react";
import T from "prop-types";
import Icon from "../Icon/component.jsx";

const DropdownMenuContext = createContext();

const DropdownMenu = ({ children }) => {
  const [isOpen, setOpen] = useState(false);
  const ref = useRef(null);

  useEffect(() => {
    const clickHandler = (e) => {
      if (ref?.current?.contains(e.target)) return;
      setOpen(false);
    };

    document.addEventListener("click", clickHandler);

    return () => {
      document.removeEventListener("click", clickHandler);
    };
  }, []);

  return (
    <DropdownMenuContext.Provider value={{ isOpen, setOpen }}>
      <div id="dropdown-menu" className="relative" ref={ref}>
        {children}
      </div>
    </DropdownMenuContext.Provider>
  );
};

DropdownMenu.propTypes = {
  children: T.node,
};

const Trigger = ({ children, additionalTriggerCSS = "" }) => {
  const { isOpen, setOpen } = useContext(DropdownMenuContext);

  return (
    <button
      id="menu-trigger"
      onClick={() => setOpen(!isOpen)}
      className={`${additionalTriggerCSS} flex items-center p-8 ml-8 group hover:text-gui-hover hover:bg-light-grey active:text-gui-active focus:text-gui-focus focus:outline-none rounded-lg`}
    >
      <span className="leading-normal">{children}</span>
      <Icon
        name="icon-gui-disclosure-arrow"
        color="text-cool-black"
        size="1.25rem"
        additionalCSS="transform rotate-90 group-hover:text-gui-hover group-active:text-gui-active group-focus:text-gui-focus"
      />
    </button>
  );
};

Trigger.propTypes = {
  children: T.node,
  additionalTriggerCSS: T.string,
};

const Content = ({ children, anchorPosition = "right", additionalContentCSS }) => {
  const { isOpen } = useContext(DropdownMenuContext);

  if (!isOpen) {
    return null;
  }

  const anchorPositionClasses = anchorPosition === "right" ? "right-0" : "left-0";

  return (
    <div
      id="menu-content"
      className={`${additionalContentCSS} absolute p-8 z-10 border border-mid-grey bg-white rounded shadow-container ${anchorPositionClasses}`}
      style={{ minWidth: 275, top: 44 }}
    >
      {children}
    </div>
  );
};

Content.propTypes = {
  children: T.node,
  anchorPosition: T.string,
  additionalContentCSS: T.string,
};

const Link = ({ url, title, subtitle, iconName, children }) => {
  return (
    <a href={url} className="menu-link group block p-8 hover:bg-light-grey hover:text-cool-black rounded">
      <p className="mb-4">
        {title}
        {iconName ? <Icon name={iconName} size="1rem" color="text-cool-black" additionalCSS="align-middle ml-8 relative -top-1 -left-4" /> : null}
      </p>
      {subtitle ? <p className="text-p3 mb-16 text-dark-grey">{subtitle}</p> : null}
      {children}
    </a>
  );
};

Link.propTypes = {
  url: T.string,
  title: T.string,
  subtitle: T.string,
  iconName: T.string,
  children: T.node,
};

DropdownMenu.Trigger = Trigger;
DropdownMenu.Content = Content;
DropdownMenu.Link = Link;
export default DropdownMenu;
