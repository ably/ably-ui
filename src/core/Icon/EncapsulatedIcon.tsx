import React from "react";
import Icon, { IconProps } from "../Icon";
import useTheming from "../hooks/useTheming";
import { Theme } from "../styles/colors/types";
import { IconSize } from "./types";

type EncapsulatedIconProps = {
  theme?: Theme;
  className?: string;
  innerClassName?: string;
  iconSize?: IconSize;
} & IconProps;

const ICON_SIZE_REDUCTION = 12;
const ICON_HEIGHT_REDUCTION = 2;

const EncapsulatedIcon = ({
  theme = "dark",
  size = "40px",
  iconSize,
  className,
  innerClassName,
  ...iconProps
}: EncapsulatedIconProps) => {
  const { themeColor } = useTheming({
    baseTheme: "dark",
    theme,
  });
  let computedIconSize = size,
    computedIconHeight = size;

  if (iconSize) {
    computedIconSize = iconSize;
  } else if (size.toString().endsWith("px")) {
    const numericalSize = parseInt(size, 10);
    computedIconSize = `${numericalSize - ICON_SIZE_REDUCTION}px`;
    computedIconHeight = `${numericalSize - ICON_HEIGHT_REDUCTION}px`;
  } else {
    console.warn(
      `EncapsulatedIcon: Non-pixel units might not behave as expected`,
    );
  }

  return (
    <div
      className={`p-1 rounded-lg ${theme === "light" ? "bg-gradient-to-t" : "bg-gradient-to-b"} ${themeColor("from-neutral-900")} ${className ?? ""}`}
      style={{ width: size, height: size }}
    >
      <div
        className={`flex items-center justify-center rounded-lg ${themeColor("bg-neutral-1100")} ${innerClassName ?? ""}`}
        style={{ height: computedIconHeight }}
      >
        <Icon size={computedIconSize} {...iconProps} />
      </div>
    </div>
  );
};

export default EncapsulatedIcon;
