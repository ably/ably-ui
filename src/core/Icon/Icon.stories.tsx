import React from "react";
import Icon from "../Icon";

export default {
  title: "Components/Icon",
  component: Icon,
  parameters: {
    layout: "fullscreen",
  },
  tags: ["autodocs"],
};

const coreIcons = [
  "icon-gui-ably-badge",
  "icon-gui-arrow-bidirectional-horizontal",
  "icon-gui-arrow-bidirectional-vertical",
  "icon-gui-arrow-down",
  "icon-gui-arrow-left",
  "icon-gui-arrow-right",
  "icon-gui-arrow-up",
  "icon-gui-burger-menu",
  "icon-gui-check-circled-fill",
  "icon-gui-check-circled-fill-black",
  "icon-gui-check-circled",
  "icon-gui-checklist-checked",
  "icon-gui-clock",
  "icon-gui-close",
  "icon-gui-cross-circled-fill",
  "icon-gui-cross-circled",
  "icon-gui-copy",
  "icon-gui-dash-circled",
  "icon-gui-disclosure-arrow",
  "icon-gui-document-generic",
  "icon-gui-enlarge",
  "icon-gui-external-link",
  "icon-gui-history",
  "icon-gui-info",
  "icon-gui-link-arrow",
  "icon-gui-live-chat",
  "icon-gui-minus",
  "icon-gui-plus",
  "icon-gui-quote-marks-solid",
  "icon-gui-refresh",
  "icon-gui-search",
  "icon-gui-tick",
  "icon-gui-warning",
  "icon-gui-link",
  "icon-gui-filter-flow-step-1",
  "icon-gui-filter-flow-step-2",
  "icon-gui-filter-flow-step-3",
  "icon-gui-resources",
];

const displayIcons = [
  "icon-display-48hrs",
  "icon-display-about-ably-col",
  "icon-display-api",
  "icon-display-api-keys",
  "icon-display-asset-tracking-col",
  "icon-display-browser",
  "icon-display-calendar",
  "icon-display-call-mobile",
  "icon-display-careers-col",
  "icon-display-case-studies-col",
  "icon-display-chat-stack-col",
  "icon-display-cloud-servers",
  "icon-display-compare-tech-col",
  "icon-display-customers-col",
  "icon-display-docs-col",
  "icon-display-documentation",
  "icon-display-examples-col",
  "icon-display-gdpr",
  "icon-display-general-comms",
  "icon-display-hipaa",
  "icon-display-it-support-access",
  "icon-display-it-support-helpdesk",
  "icon-display-laptop",
  "icon-display-lightbulb-col",
  "icon-display-live-chat",
  "icon-display-map-pin",
  "icon-display-message",
  "icon-display-padlock-closed",
  "icon-display-platform",
  "icon-display-play",
  "icon-display-privacy-shield-framework",
  "icon-display-quickstart-guides-col",
  "icon-display-resources-col",
  "icon-display-sdks-col",
  "icon-display-servers",
  "icon-display-shopping-cart",
  "icon-display-sla",
  "icon-display-chat-col",
  "icon-display-soc2-type2",
  "icon-display-tech-account-comms",
  "icon-display-tutorials-demos-col",
  "icon-display-integrations-col",
  "icon-display-virtual-events-col",
  "icon-live-updates-results-metrics-col",
  "icon-multi-user-spaces-col",
];

const socialIcons = [
  "discord",
  "facebook",
  "github",
  "glassdoor",
  "linkedin",
  "twitter",
  "google",
  "stackoverflow",
  "youtube",
];

const otherIcons = ["quote"];

const renderIcons = (iconSet: string[]) => (
  <div className="grid grid-cols-4 ui-grid-gap max-w-screen-lg mb-64">
    {iconSet.map((icon: string) => (
      <div className="border p-16" key={icon}>
        <code className="ui-text-code mb-8 block">{icon}</code>
        <div className="border inline-flex flex-0">
          <div className="flex pi-checkered-bg">
            <Icon
              name={icon}
              additionalCSS="hover:text-active-orange"
              color="text-cool-black"
              size="1.5rem"
            />
          </div>
        </div>
      </div>
    ))}
  </div>
);

export const GUIIcons = {
  render: () => renderIcons(coreIcons),
};

export const DisplayIcons = {
  render: () => renderIcons(displayIcons),
};

export const SocialIcons = {
  render: () => renderIcons(socialIcons),
};

export const Other = {
  render: () => renderIcons(otherIcons),
};
