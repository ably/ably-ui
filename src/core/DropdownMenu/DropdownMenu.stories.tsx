import React from "react";
import Icon from "../Icon/component.jsx";
import DropdownMenu from "./component.jsx";
import { StoryObj } from "@storybook/react";

export default {
  title: "Components/Dropdown Menu",
  component: DropdownMenu,
};

type Story = StoryObj<typeof DropdownMenu>;

export const Default: Story = {
  render: () => (
    <DropdownMenu>
      <DropdownMenu.Trigger additionalTriggerCSS="ui-text-h3">
        Dropdown Menu Trigger
      </DropdownMenu.Trigger>
      <DropdownMenu.Content anchorPosition="left">
        <DropdownMenu.Link
          url="https://ably.com/"
          title="Using DropdownMenuLink Component"
          subtitle="This is using the inbuilt component which takes props for title, subtitle, icon name & children."
          iconName="icon-gui-link-arrow"
        >
          <p className="ui-text-p3">I am a child! üê£</p>
        </DropdownMenu.Link>
        <DropdownMenu.Link
          url="https://ably.com/"
          title="Using DropdownMenuLink Component - no icon, subtitle or children"
        />
        <a
          href="https://ably.com/docs"
          target="_blank"
          className="group block p-8 hover:bg-light-grey hover:text-cool-black rounded"
        >
          <p className="ui-featured-link group-hover:text-gui-hover font-light text-cool-black">
            Using plain HTML
            <Icon
              name="icon-gui-link-arrow"
              size="1rem"
              color="text-cool-black"
              additionalCSS="ui-featured-link-icon group-hover:text-gui-hover group-active:text-gui-active group-focus:text-gui-focus"
            />
          </p>
        </a>
      </DropdownMenu.Content>
    </DropdownMenu>
  ),
};
