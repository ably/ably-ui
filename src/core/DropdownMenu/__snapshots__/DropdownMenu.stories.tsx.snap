// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Dropdown Menu Default smoke-test 1`] = `
<div id="dropdown-menu"
     class="relative"
>
  <button id="menu-trigger"
          role="button"
          class="flex items-center ui-text-label3 font-bold text-neutral-1000 dark:text-neutral-300 focus:outline-none ui-text-h3"
  >
    Dropdown Menu Trigger
    <svg class="inline-block text-neutral-1300 dark:text-neutral-000"
         style="width: 1.25rem; height: 1.25rem;"
    >
      <use xlink:href="#sprite-icon-gui-chevron-down-mini">
      </use>
    </svg>
  </button>
</div>
`;
